pkg/route/api/validation/validation.go

```diff
@@ -102,9 +108,121 @@ func validateTLS(tls *routeapi.TLSConfig) fielderrors.ValidationErrorList {
 		if len(tls.DestinationCACertificate) > 0 {
 			result = append(result, fielderrors.NewFieldInvalid("destinationCACertificate", tls.DestinationCACertificate, "edge termination does not support destination certificates"))
 		}
+
+		result = append(result, validateTLSCertificates(tls)...)
+
 	default:
 		msg := fmt.Sprintf("invalid value for termination, acceptable values are %s, %s, %s, or emtpy (no tls specified)", routeapi.TLSTerminationEdge, routeapi.TLSTerminationPassthrough, routeapi.TLSTerminationReencrypt)
 		result = append(result, fielderrors.NewFieldInvalid("termination", tls.Termination, msg))
 	}
 	return result
 }
+
+// validateTLSCertificates checks that the certificates passed are able to be parsed.  This includes ensuring that
+// the cert/key pair are matching and, if a ca cert is provided it will check the first ca cert against the leaf
+// to ensure that the leaf was signed by the ca
+func validateTLSCertificates(tls *routeapi.TLSConfig) fielderrors.ValidationErrorList {
+	if tls == nil {
+		return nil
+	}
+	//edge and reencrypt (the use cases that call this require all 3 (at least) so we can ignore anything
+	//that doesn't have all 3 set.  It will already fail validation elsewhere
+	if len(tls.Certificate) == 0 || len(tls.Key) == 0 || len(tls.CACertificate) == 0 {
+		return nil
+	}
+
+	result := fielderrors.ValidationErrorList{}
+
+	//check cert/key pair
+	checkSignature := true
+	cert, err := ctls.X509KeyPair([]byte(tls.Certificate), []byte(tls.Key))
```

Comment:
We can always require users to give us certs in PEM format and have openssl in the router do the external conversion (in the short term, if we limited to PEM).

As is I'm fine with this continuing as long as we document in lots of places that we require PEM certificates and the errors are clear (i.e. "expected PEM certificate format: %v")

----- Original Message -----
> > +// to ensure that the leaf was signed by the ca
> > +func validateTLSCertificates(tls *routeapi.TLSConfig)
> > fielderrors.ValidationErrorList {
> > +	if tls == nil {
> > +		return nil
> > +	}
> > +	//edge and reencrypt (the use cases that call this require all 3 (at
> > least) so we can ignore anything
> > +	//that doesn't have all 3 set.  It will already fail validation elsewhere
> > +	if len(tls.Certificate) == 0 || len(tls.Key) == 0 ||
> > len(tls.CACertificate) == 0 {
> > +		return nil
> > +	}
> > +
> > +	result := fielderrors.ValidationErrorList{}
> > +
> > +	//check cert/key pair
> > +	checkSignature := true
> > +	cert, err := ctls.X509KeyPair([]byte(tls.Certificate), []byte(tls.Key))
> 
> I think you meant we can require PEM.
> 
> Since we know we want to support other formats in the future I think we
> should investigate golang wrappers for openssl.  If openssl can already
> load/validate all the formats for us and do conversion it seems like the
> best choice.  I'll mark this as WIP while we research.
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/1824/files#r28797142